driver-core/src/main/com/mongodb/internal/connection/DefaultServer.java

```diff
@@ -126,52 +127,46 @@ public ServerDescription getDescription() {
     }
 
     @Override
-    public void invalidate() {
-        invalidate(connectionPool.getGeneration());
+    public void resetToConnecting() {
+        serverStateListener.stateChanged(new ChangeEvent<>(description, ServerDescription.builder()
+                .state(CONNECTING).address(serverId.getAddress()).build()));
     }
 
     @Override
-    public void invalidate(final int connectionGeneration) {
+    public void invalidate() {
         if (!isClosed()) {
-            if (connectionGeneration < connectionPool.getGeneration()) {
-                return;
-            }
-            serverStateListener.stateChanged(new ChangeEvent<ServerDescription>(description, ServerDescription.builder()
-                    .state(CONNECTING)
-                    .address(serverId.getAddress())
-                    .build()));
-            connectionPool.invalidate();
-            // TODO: in streaming protocol, we want to close the current connection and start over
+            serverStateListener.stateChanged(new ChangeEvent<>(description, ServerDescription.builder()
+                    .state(CONNECTING).address(serverId.getAddress()).build()));
             connect();
+            if (description.getMaxWireVersion() < FOUR_DOT_TWO_WIRE_VERSION) {
+                connectionPool.invalidate();
+            }
         }
     }
 
     @Override
-    public void invalidate(final Throwable reason) {
-        invalidate(reason, connectionPool.getGeneration(), description.getMaxWireVersion());
-    }
-
-    @Override
-    public void invalidate(final Throwable t, final int connectionGeneration, final int maxWireVersion) {
+    public void invalidate(final ConnectionState connectionState, final Throwable t, final int connectionGeneration,
```

Comment:
By adding ConnectionState, we can share this method for both the before and after handshake behaviors, as there is only one small difference between them.